### PR TITLE
Ensure no public type lies under Core namespace

### PR DESCRIPTION
**Pull request by nulltoken**

### CODE REVIEW

**Comment by nulltoken** [comment]
This fails because we've still got Epoch under the `Core` namespace.

What should we do with it?

As it's mostly used by the tests, we could make it internal and share (ie. link) the file between the two projects.

**Comment by nulltoken** [comment]
>As it's mostly used by the tests, we could make it internal and share (ie. link) the file between the two projects.

I've just pushed a proposal for this.

**Comment by nulltoken** [comment]
Ready for review

**Comment by nulltoken** [comment]
@Therzok Hmmm. Looks like Mono `Assembly.GetExportedTypes` doesn't match .Net one

```
	/home/travis/build/libgit2/libgit2sharp/CI/build.msbuild: error : LibGit2Sharp.Tests.MetaFixture.NoPublicTypesUnderLibGit2SharpCoreNamespace: 
Public type 'LibGit2Sharp.Core.GitSmartSubtransportRegistration+create_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.Proxy+Func`6' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitSmartSubtransport+action_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitSmartSubtransport+close_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitSmartSubtransport+free_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitSmartSubtransportStream+read_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitSmartSubtransportStream+write_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitSmartSubtransportStream+free_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackend+read_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackend+read_prefix_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackend+read_header_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackend+write_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackend+writestream_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackend+readstream_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackend+exists_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackend+exists_prefix_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackend+foreach_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackend+free_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackend+foreach_callback_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackendStream+read_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackendStream+write_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackendStream+finalize_write_callback' under the 'LibGit2Sharp.Core' namespace.
Public type 'LibGit2Sharp.Core.GitOdbBackendStream+free_callback' under the 'LibGit2Sharp.Core' namespace.
```

File: LibGit2Sharp.Tests/MetaFixture.cs [hunk 1]
```diff
@@ -254,12 +254,49 @@ public void GetEnumeratorMethodsInLibGit2SharpMustBeVirtualForTestability()
...
+        {
+            const string coreNamespace = "LibGit2Sharp.Core";
+
+            var types = Assembly.GetAssembly(typeof(IRepository))
+                .GetExportedTypes()
+                .Where(t => t.FullName.StartsWith(coreNamespace + "."))
 
-            Assert.Empty(nonVirtualGetEnumeratorMethods);
+                // Ugly hack to circumvent a Mono bug
+                .Where(t => !t.FullName.Contains("+"))
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/MetaFixture.cs]: Eww..

Comment (nulltoken) [inline LibGit2Sharp.Tests/MetaFixture.cs]: Is it "safe" enough as a hack?

Comment (Therzok) [inline LibGit2Sharp.Tests/MetaFixture.cs]: You can't have "+" in names, so yeah. I'd say it's safe for now.

Comment (nulltoken) [inline LibGit2Sharp.Tests/MetaFixture.cs]: >You can't have "+" in names, so yeah. I'd say it's safe for now.

That was my understanding as well. Thanks for the feedback!

BTW is this a known issue?

Comment (Therzok) [inline LibGit2Sharp.Tests/MetaFixture.cs]: I'm opening a bug report now.

Comment (Therzok) [inline LibGit2Sharp.Tests/MetaFixture.cs]: Actually, I don't know whether it's a runtime/bcl/compiler bug. Asking around first.

Comment (nulltoken) [inline LibGit2Sharp.Tests/MetaFixture.cs]: Thanks for this. I'll update the PR comment with the link to the bug entry

**Comment by Therzok** [comment]
https://bugzilla.xamarin.com/show_bug.cgi?id=27010

**Comment by nulltoken** [comment]
@Therzok :heart:

**Comment by Therzok** [comment]
Fix was merged for this. It should probably land in the next release of Mono.

https://github.com/mono/mono/pull/1570#event-237185788

**Pull request merged**